zerver/migrations/0260_missed_message_addresses_from_redis_to_db.py

```diff
@@ -0,0 +1,92 @@
+from django.db.migrations.state import StateApps
+
+from zerver.lib.redis_utils import get_redis_client
+from zerver.lib.utils import generate_random_token
+
+def generate_missed_message_token() -> str:
+    return 'mm' + generate_random_token(32)
+
+def move_missed_message_addresses_to_database(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    redis_client = get_redis_client()
+    MissedMessageEmailAddress = apps.get_model('zerver', 'MissedMessageEmailAddress')
+    UserProfile = apps.get_model('zerver', 'UserProfile')
+    Message = apps.get_model('zerver', 'Message')
+    Recipient = apps.get_model('zerver', 'Recipient')
+    RECIPIENT_PERSONAL = 1
+    RECIPIENT_STREAM = 2
+
+    all_mm_keys = redis_client.keys('missed_message:*')
+    for key in all_mm_keys:
+        # Don't migrate mm addresses that have already been used.
+        if redis_client.hincrby(key, 'uses_left', -1) < 0:
+            redis_client.delete(key)
+            continue
+
+        result = redis_client.hmget(key, 'user_profile_id', 'recipient_id', 'subject')
+        if not all(val is not None for val in result):
+            # Missing data, skip this key
+            redis_client.delete(key)
+            continue
+
+        user_profile_id, recipient_id, subject_b = result  # type: (bytes, bytes, bytes)
+        topic_name = subject_b.decode('utf-8')
+
+        # Now we have to figure out the appropriate objects to which ForeignKey fields
+        # of MissedMessageEmailAddress should point. UserProfile and Recipient are straightforward,
+        # the main complication is with Message, since the redis data doesn't hold information
+        # about which Message exactly the "missed message" actually was. We take a "best effort"
+        # approach here and select the "latest most viable" Message. For personals and huddles that's simple,
+        # for streams we try "latest message with the topic name" and fall back to "latest message in the stream"
+        # if the topic no longer exists.
+        # If the necessary objects can't be found in the database, we will catch ObjectDoesNotExist and skip this address.
+        try:
+            user_profile = UserProfile.objects.get(id=user_profile_id)
+            recipient = Recipient.objects.get(id=recipient_id)
+
+            if recipient.type == RECIPIENT_STREAM:
+                try:
+                    message = Message.objects.filter(subject__iexact=topic_name,
+                                                     recipient_id=recipient.id).latest('id')
+                except Message.DoesNotExist:
+                    # The topic no longer exists or has been renamed, so we have to make a compromise here.
+                    # We'll tie this missed message address to the latest message in the stream.
+                    # (Or raise DoesNotExist to be caught, if even that's not possible.)
+                    message = Message.objects.filter(recipient_id=recipient.id).latest('id')
```

Comment:
I think it might be safer in corner cases where we can't find a suitable message to associate to just not create an entry in this migration; the only harm will be that email replies to those messages will get an error over that 5-day window.

Reply:
(Which I think means we just remove this except block?)

Reply:
I got rid of the block